### ISSUE TITLE
View full exception & stacktrace in Node.JS debugging

### ISSUE THREAD

**Issue by SqrtMinusOne** [emacs-lsp/dap-mode#416]
When debugger pauses at an exception, message & stack are shown in the locals view. The problem is that they are cropped, as seen on the screenshot, and I can't figure out how to display the message in full.

![image](https://user-images.githubusercontent.com/21692287/100520473-e1f60780-31ae-11eb-9e26-bf7f0e035cee.png)

I figured out I can use REPL to display long variables, but there is no way I know to access an ongoing exception from the REPL.

For instance, here's how Chrome DevTools handle this:
![image](https://user-images.githubusercontent.com/21692287/100520748-89c00500-31b0-11eb-98a3-25e0d8a6cc5a.png)

The message can be seen in full in the pane and in the tooltip, and it also can be saved as a global variable to be inspected in the REPL.

Is there a way to replicate something like that in dap-mode?

- `dap-mode` version: 0.6 at the commit e582ff90c104703ed65c1f3174d4c4690e9cf1fd
- Emacs version: 27.1

**Comment by yyoncho** [comment]
We should have hover for the truncated messages. You may change dap-ui-variable-length  which will affect how we truncate the data. About the exception message, it should be present in the messages buffer, no? The open question I think is how to copy the value of truncated element.

**Comment by SqrtMinusOne** [comment]
Thanks, changing `dap-ui-variable-length` may be just enough, as long lines are truncated in the view anyhow. Copying the value from the locals view would be useful though.

As for the exception message, at least in Node.js, the exception message may appear in the output buffer when the error handling flow is over, that is when exception is uncaught or was printed in a catch block. Whereas the debugger pauses the execution right at the start of the error handling flow, before the message is printed.

And tooltips don't seem to work in my case. A tooltip does appear in the code buffer when I call `dap-tooltip-at-point`, but it doesn't appear in the locals buffer. The problem maybe with my setup, but I have just the following for lsp-mode:
```emacs-lisp
(use-package lsp-mode
  :straight t
  :hook (
         (typescript-mode . lsp)
         (vue-mode . lsp)
         (go-mode . lsp)
         (svelte-mode . lsp)
         (python-mode . lsp)) 
  :commands lsp
  :config
  (setq lsp-idle-delay 1)
  (setq lsp-signature-render-documentation nil)
  (add-to-list 'lsp-language-id-configuration '(svelte-mode . "svelte"))  )
  
(use-package lsp-ui
  :straight t
  :commands lsp-ui-mode
  :config
  (setq lsp-ui-doc-delay 2)
  (setq lsp-ui-sideline-show-hover nil))

(use-package lsp-treemacs
  :straight t
  :commands lsp-treemacs-errors-list)
```
And the following for dap-mode:
```emacs-lisp
(use-package dap-mode
  :straight t
  :init
  (setq lsp-enable-dap-auto-configure nil)
  :config

  (require 'dap-node)
  (dap-node-setup)

  (require 'dap-chrome)
  (dap-chrome-setup)
  
  (dap-mode 1)
  (dap-ui-mode 1)
  (dap-tooltip-mode 1)
  (tooltip-mode 1)
  (dap-ui-controls-mode 1))

```

**Comment by nbfalcon** [comment]
@yyoncho something like
`(plist-get (button-get (treemacs-node-at-point) :item) :value)` would get the value, untruncated.

@SqrtMinusOne
This function will yank the value under the cursor to the kill ring. Don't close this yet, as we need this in `master`.
```emacs-lisp
(defun my/dap-yank-value-at-point (node)
  (interactive (list (treemacs-node-at-point)))
  (kill-new (plist-get (button-get node :item) :value)))
```

**Comment by yyoncho** [comment]
> And tooltips don't seem to work in my case. A tooltip does appear in the code buffer when I call `dap-tooltip-at-point`, but it doesn't appear in the locals buffer. The problem maybe with my setup, but I have just the following for lsp-mode:

We are setting 'help-echo for the string, which should get you the content of the value when doing mouse hover. 

> As for the exception message, at least in Node.js, the exception message may appear in the output buffer when the error handling flow is over, that is when exception is uncaught or was printed in a catch block. Whereas the debugger pauses the execution right at the start of the error handling flow, before the message is printed.

The error message should be present in the `*Messages*` buffer, not in the output buffer, at least for Caught/Uncaught errors.

**Comment by nbfalcon** [comment]
Yeah, the options are either node -> `:item` -> `:value` or `get-text-property (point) 'help-echo`.

**Comment by yyoncho** [comment]
@nbfalcon I think we should improve the exception breakpoint related UI... We might add an entry in the breakpoints view. ATM we only do `lsp--error` which seems to be less convenient.

**Comment by SqrtMinusOne** [comment]
@nbfalcon That's awesome, thanks!

@yyoncho There is indeed a message in the `*Messages*`, but not quite what I expected:
```
LSP :: Exception has occurred: Error
nil
```

Also, hover message does not appear in my case, but `(get-text-property (point) 'help-echo)` does indeed return help text. The problem is that when I click on a node, the cursor apparently positions at the beginning of the line, where `help-echo` is `nil`. After I move cursor right a few times, `help-echo` appears. Which is not too obvious since the cursor is invisible there.
![image](https://user-images.githubusercontent.com/21692287/100543204-d0206d00-325f-11eb-8143-a80d8282ff5a.png)

That's probably not a big deal, since there is another way to access the full value without such limitations.

**Comment by nbfalcon** [comment]
@yyoncho I agree that `dap-mode` still needs to be improved a lot (`lsp--error` for Exceptions is wrong), and I'll be working on doing so soon. However, currently I am a bit choked for time. We should consider using `eldoc` here.

@SqrtMinusOne `get-text-property` should move `point`; you can add my `interactive` function to your config while we don't have that in feature in `master`. I have edited my original answer, so that the function I posted now writes a message to the echo area. Position your cursor on the exception, and then type `M-x my/dap-yank-value-at-point`.
